cli/common.ml

```diff
@@ -43,3 +60,37 @@ module Logs = struct
     Logs.app ?src (fun l ->
         f (fun ?header ?tags fmt -> l ?header ?tags ("%a" ^^ fmt) Duniverse_lib.Styled_pp.header ()))
 end
+
+(** Checks that all provided repos match a source repo in the duniverse *)
+let check_duniverse_repos duniverse_repos src_deps =
+  let dir_names = List.map ~f:(fun (s : _ Duniverse.Deps.Source.t) -> s.dir) src_deps in
+  let found = List.sort_uniq ~compare:String.compare dir_names in
+  let asked = List.sort_uniq ~compare:String.compare duniverse_repos in
+  let rec diff acc asked_l found_l =
+    match (asked_l, found_l) with
+    | asked :: asked_tl, found :: found_tl ->
+        if String.equal asked found then diff acc asked_tl found_tl
+        else diff (asked :: acc) asked_tl found_l
+    | asked :: asked_tl, [] -> diff (asked :: acc) asked_tl found_l
+    | [], [] -> List.rev acc
+    | [], _ -> assert false
+  in
+  match diff [] asked found with
+  | [] -> Ok ()
+  | unmatched ->
+      let sep fmt () = Ffmt.pf fmt " " in
+      Rresult.R.error_msgf "The following repos are not in your duniverse: %a"
+        Ffmt.(list ~sep string)
+        unmatched
+
+let should_consider ~to_consider (src_dep : _ Duniverse.Deps.Source.t) =
+  List.mem ~set:to_consider src_dep.dir
+
+(** Filters the duniverse according to the CLI provided list of repos *)
+let filter_duniverse ~to_consider src_deps =
+  let open Rresult in
+  match to_consider with
+  | None -> Ok src_deps
+  | Some to_consider ->
+      let filtered = List.filter ~f:(should_consider ~to_consider) src_deps in
+      check_duniverse_repos to_consider filtered >>= fun () -> Ok filtered
```

Comment:
This is a lot of code ! I think all this can be simplified with an assumptions: the result correspond to `to_consider` (it's not filtered). The only needed logic is to lookup in `src_deps` (as a map):
```ocaml

(** Filters the duniverse according to the CLI provided list of repos *)
let filter_duniverse ~to_consider src_deps =
  let open Rresult in
  match to_consider with
  | None -> Ok src_deps
  | Some to_consider -> (
      let found =
        let f (src : _ Duniverse.Deps.Source.t) = (src.dir, src) in
        String.Map.of_list_map_exn src_deps ~f
      in
      let unmatched, to_consider =
        List.partition_map to_consider ~f:(fun asked ->
            match String.Map.find found asked with None -> Left asked | Some src -> Right src)
      in
      match (unmatched, to_consider) with
      | [], to_consider -> Ok to_consider
      | unmatched, _ ->
          let sep fmt () = Ffmt.pf fmt " " in
          Rresult.R.error_msgf "The following repos are not in your duniverse: %a"
            Ffmt.(list ~sep string)
            unmatched )
```

This example uses functions from stdune. (which is in scope because of dune-configurator and used in every modules)

Reply:
You're right this looks much better! I'm not sure what I had in mind when I wrote that :')